readme.md
# Hal
 * [HAL Specification](http://stateless.co/hal_specification.html)
 * [Hal Specification on Github](https://github.com/mikekelly/hal_specification)
 * [JSON Linking With HAL](http://blog.stateless.co/post/13296666138/json-linking-with-hal)
 * [Linking In Json](http://www.mnot.net/blog/2011/11/25/linking_in_json)
 * [Examples of HAL](https://gist.github.com/2289546)
 * [HAL on Google Groups](https://groups.google.com/d/forum/hal-discuss)

```php
<?php
/* Create a new Resource Parent */
$parent = new Hal_Resource('/dogs');
/* Add any relevent links */
$parent->setLink(new Hal_Link('/dogs?q={text}', 'search'));
$dogs[1] =  new Hal_Resource('/dogs/1');
$dogs[1]->setData(
    array(
        'id' => '1', 
        'name' => 'tiber', 
        'color' => 'black'
    )
);
$dogs[2] =  new Hal_Resource(
    '/dogs/2',array(
        'id' => '2', 
        'name' => 'sally', 
        'color' => 'white'
    )
);
$dogs[3] =  new Hal_Resource(
    '/dogs/3',array(
        'id' => '3', 
        'name' => 'fido', 
        'color' => 'gray'
    )
);
/* Add the embedded resources */
foreach ($dogs as $dog) {
    $parent->setEmbedded('dog', $dog);
}
echo (string) $parent;
```

### Result: 

```javascript
{
  "_links":{
		"self":{
			"href":"\/dogs"
		},
		"search":{
			"href":"\/dogs?q={text}"
		}
	},
	"_embedded":{
		"dog":[
			{
				"_links":{
					"self":{
						"href":"\/dogs\/1"
					}
				},
				"id":"1",
				"name":"tiber",
				"color":"black"
			},
			{
				"_links":{
					"self":{
						"href":"\/dogs\/2"
					}
				},
				"id":"2",
				"name":"sally",
				"color":"white"
			},
			{
				"_links":{
					"self":{
						"href":"\/dogs\/3"
					}
				},
				"id":"3",
				"name":"fido",
				"color":"gray"
			}
		]
	}
}
```
## Generating XML output

```php
<?php
echo $parent->getXML()->asXML();
```
### Result:
```xml
<?xml version="1.0"?>
<resource href="/dogs">
    <link href="/dogs?q={text}" rel="search" />
    <resource href="/dogs/1" rel="dog">
        <id>1</id>
        <name>tiber</name>
        <color>black</color>
    </resource>
    <resource href="/dogs/2" rel="dog">
        <id>2</id>
        <name>sally</name>
        <color>white</color>
    </resource>
    <resource href="/dogs/3" rel="dog">
        <id>3</id>
        <name>fido</name>
        <color>gray</color>
    </resource>
</resource>
```


CHANGELOG
commit be63763f8e22ae75ce79289aee8631ea6d9a5b89
Merge: 8dc2a54 872151c
Author: Robert Allen <zircote@gmail.com>
Date:   Fri Apr 6 16:35:24 2012 -0700

    Merge pull request #1 from hradtke/links
    
    Allow multiple href's per link relation.

commit 872151ced307914989b39e6a59743980098fa024
Author: Herman J. Radtke III <hermanradtke@gmail.com>
Date:   Fri Apr 6 15:25:45 2012 -0700

    Allow multiple href's per link relation.

commit 8dc2a542de66c7e770e00d321f4784cf26aa90e5
Author: Robert Allen <zircote@gmail.com>
Date:   Fri Apr 6 09:25:28 2012 -0500

    Added link to hal spec repository

commit 8816b87b42c8e84f427624694ca3b92a14c12fe2
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Apr 5 21:57:24 2012 -0500

    Updated Links and readme

commit a518dbbbaa295454db577c4670f4b8b996743079
Merge: 33e2a43 a66640f
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Apr 5 13:30:59 2012 -0500

    merged changes from @hradtke resolved conflicts
    detabbed

commit 33e2a43bd5761a3d9a52a39fb039b0880a21b127
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Apr 5 12:44:25 2012 -0500

    minors editing of readme for readability sake

commit 926ae6fcb1f40ea91d02b96ee4547a4fff106fb2
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Apr 5 12:41:38 2012 -0500

    removed the ^Ms as line delimters -.-

commit 8ac4bf6fb59d46796517b65943eba1067cb1e0bf
Merge: a1e39fb 1eddd70
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Apr 5 12:35:49 2012 -0500

    Merge branch 'develop'

commit a1e39fb0f022dfc38eaaff2e2959a855e10e8aa7
Merge: 3fe4d10 fb3f375
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Apr 5 12:34:25 2012 -0500

    refactoring of namespace, and incorporated the changes made by @hradtke

commit 1eddd701c79b3bf7929563a65b20b30aeb6624a4
Merge: 3fe4d10 fb3f375
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Apr 5 12:34:25 2012 -0500

    Merge branch 'master' into develop

commit fb3f375dc9313fc3d64b7183be71b9e114020800
Merge: c9e301e bc59106
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Apr 5 12:31:55 2012 -0500

    refactoring of namespace, and incorporated the changes made by @hradtke

commit 3fe4d1012fdc316a17d9c128cfcd5b65c70dbec0
Merge: c9e301e bc59106
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Apr 5 12:31:55 2012 -0500

    Merge branch 'master' into develop

commit bc59106e0af0ea14ce71d1c8aaef89b43aecf569
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Apr 5 12:29:51 2012 -0500

    refactoring of namespace, and incorporated the changes made by @hradtke

commit c9e301e64c8c1faf820be60970cd53ad9ee121ed
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Apr 5 12:29:51 2012 -0500

    refactoring of namespace, and incorporated the changes made by @hradtke

commit a66640f469fe7ec3a625c708a434f12d40b6efa1
Author: Herman J. Radtke III <hermanradtke@gmail.com>
Date:   Wed Mar 28 17:29:50 2012 -0700

    Fixed the Hal_Resource interface and fixed format of code.
    
    The Hal_Resource contstructor was asking for a rel tag and never used it. A
    more common use case is to set the data in the constructor so the client does
    not have to call the setData() method after the fact.

commit 9ba56231b1eb2ea203eb4dd71c1cb726c5353bb3
Author: Herman J. Radtke III <hermanradtke@gmail.com>
Date:   Wed Mar 28 11:59:13 2012 -0700

    Remove Zircote namespace and Zend Framework dependency.

commit 5472967c41be0bf4b62e60726b71616713010a9f
Author: Robert Allen <zircote@gmail.com>
Date:   Fri Mar 9 17:55:21 2012 -0600

    corrected bug with data entity general

commit 652a6f764c7d6b58f18ca096fa572ba816ad72f1
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Mar 8 09:21:44 2012 -0600

    added license information

commit c99ae924e4037295a929de77f22416621cf521da
Author: Robert Allen <zircote@gmail.com>
Date:   Thu Mar 8 09:17:09 2012 -0600

    added Apache 2.0 License

commit 5338ce8c098309901a2844990926798273d933b2
Author: Robert Allen <zircote@gmail.com>
Date:   Fri Mar 2 17:52:10 2012 -0600

    updated docs, added test for simple

commit ee23c867614171c7823c430e52368775165416d6
Author: Robert Allen <zircote@gmail.com>
Date:   Fri Mar 2 16:01:45 2012 -0600

    Added XML serialization support.

commit 293eab5cc9e4036850179a385f9d8bc1e10dd40d
Author: zircote <zircote@gmail.com>
Date:   Wed Feb 29 01:16:38 2012 -0600

    Updated docblox and added relevant comments as stated in the spec.
    Refactored classes removed unnecasary Abstract.
    Change-Id: I3427b012ba52504ec44539c7e8eca32c47b54543

commit 204a37e0dae483e40ab0f207ecbaf2a02f159fa8
Author: Robert Allen <zircote@gmail.com>
Date:   Tue Feb 28 10:31:43 2012 -0600

    added links

commit a0b32a419b95f54a67a4d6d3058d195ed96318fa
Author: Robert Allen <zircote@gmail.com>
Date:   Tue Feb 28 10:21:58 2012 -0600

    Added examples and links

commit 01e9534cfd17271b02c2a7ed5945cd05875e597e
Author: Robert Allen <zircote@gmail.com>
Date:   Tue Feb 28 10:00:14 2012 -0600

    initial commit


tests/bootstrap.php
<?php
/**
 *
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * Copyright [2012] [Robert Allen]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
// Define path to application directory
defined('APPLICATION_PATH')
    || define('APPLICATION_PATH', realpath(dirname(__FILE__) . '/../'));
// Define application environment
defined('APPLICATION_ENV')
    || define('APPLICATION_ENV', (getenv('APPLICATION_ENV') ? getenv('APPLICATION_ENV') : 'testing'));

// Ensure library/ is on include_path
set_include_path(implode(PATH_SEPARATOR, array(
    realpath(APPLICATION_PATH . '/library'),
    get_include_path(),
)));

require_once 'Hal/AbstractHal.php';
require_once 'Hal/Resource.php';
require_once 'Hal/Link.php';


tests/library/Hal/LinkTest.php
<?php
/**
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * Copyright [2012] [Robert Allen]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Hal_Link test case.
 */
class Hal_LinkTest extends PHPUnit_Framework_TestCase
{
    /**
     * Prepares the environment before running a test.
     */
    protected function setUp ()
    {
        parent::setUp();
    }
    /**
     * Cleans up the environment after running a test.
     */
    protected function tearDown ()
    {
        parent::tearDown();
    }
    public function testXML()
    {
        $parentRes = new Hal_Resource('/orders');
        $parentRes->setLink(
        new Hal_Link('/orders?page=2', 'next')
        )->setLink(
        new Hal_Link('/orders?id={order_id}', 'search')
        );
        $data = array(
        'total' => '30.00','currency' => 'USD',
        'status' => 'shipped','placed' => '2011-01-16',
        );
        $embedded1 = new Hal_Resource('/orders/123');
        $embedded1->setData($data)
        ->setLink(
        new Hal_Link(
        '/customer/bob','customer','Bob Jones <bob@jones.com>'
        )
        );
        $basketItems = array(
        array('sku' => 'ABC123','quantity' => 2,'price' => '9.50'),
        array('sku' => 'GFZ111','quantity' => 1,'price' => '11.00')
        );
        $basket = new Hal_Resource('/orders/123/basket');
        $basket->setData('items', $basketItems);
        $embedded1->setEmbedded('basket', $basket, true);
        //////////////////////////////////////////////
        $data = array(
        'total' => '20.00','currency' => 'USD',
        'status' => 'processing','placed' => '2011-01-16',
        );
        $embedded2 = new Hal_Resource('/orders/124');
        $embedded2->setData($data)
        ->setLink(
        new Hal_Link(
        '/customer/jen','customer','Jen Harris <jen@internet.com>'
        )
        );
        /////////////////////////////////////
        $basketItems = array(
        array('sku' => 'KLM222','quantity' => 1,'price' => "9.00"),
        array('sku' => 'HHI50','quantity' => 1,'price' => "11.00")
        );
        $basket2 = new Hal_Resource('/orders/124/basket');
        $basket2->setData('items', $basketItems);
        $embedded2->setEmbedded('basket', $basket2, true);

        $parentRes
        ->setEmbedded('order', $embedded1)
        ->setEmbedded('order', $embedded2);
        /* @export as XML
         *
         *
         */
        $parentRes->getXML()->asXML();

    }
    public function testBuild()
    {
        $parentRes = new Hal_Resource('/orders');
        $parentRes->setLink(
            new Hal_Link('/orders?page=2', 'next')
        )->setLink(
            new Hal_Link('/orders?id={order_id}', 'search')
        );
        $data = array(
            'total' => '30.00','currency' => 'USD',
            'status' => 'shipped','placed' => '2011-01-16',
        );
        $embedded1 = new Hal_Resource('/orders/123');
        $embedded1->setData($data)
            ->setLink(
                new Hal_Link(
                    '/customer/bob','customer','Bob Jones <bob@jones.com>'
                )
            );
        $basketItems = array(
            array('sku' => 'ABC123','quantity' => 2,'price' => '9.50'),
            array('sku' => 'GFZ111','quantity' => 1,'price' => '11.00')
        );
        $basket = new Hal_Resource('/orders/123/basket');
        $basket->setData('items', $basketItems);
        $embedded1->setEmbedded('basket', $basket, true);
        //////////////////////////////////////////////
        $data = array(
            'total' => '20.00','currency' => 'USD',
            'status' => 'processing','placed' => '2011-01-16',
        );
        $embedded2 = new Hal_Resource('/orders/124');
        $embedded2->setData($data)
        ->setLink(
            new Hal_Link(
                '/customer/jen','customer','Jen Harris <jen@internet.com>'
            )
        );
        /////////////////////////////////////
        $basketItems = array(
            array('sku' => 'KLM222','quantity' => 1,'price' => "9.00"),
            array('sku' => 'HHI50','quantity' => 1,'price' => "11.00")
        );
        $basket2 = new Hal_Resource('/orders/124/basket');
        $basket2->setData('items', $basketItems);
        $embedded2->setEmbedded('basket', $basket2, true);

        $parentRes
        ->setEmbedded('order', $embedded1)
        ->setEmbedded('order', $embedded2);
        $actual = json_decode($parentRes);
        $JSON =<<<EOF
{
  "_links": {
    "self": { "href": "/orders" },
    "next": { "href": "/orders?page=2" },
    "search": { "href": "/orders?id={order_id}" }
  },
  "_embedded": {
    "order": [
      {
        "_links": {
          "self": { "href": "/orders/123" },
          "customer": { "href": "/customer/bob", "title": "Bob Jones <bob@jones.com>" }
        },
        "total": "30.00",
        "currency": "USD",
        "status": "shipped",
        "placed": "2011-01-16",
        "_embedded": {
          "basket": {
            "_links": {
              "self": { "href": "/orders/123/basket" }
            },
            "items": [
              {
                "sku": "ABC123",
                "quantity": 2,
                "price": "9.50"
              },{
                "sku": "GFZ111",
                "quantity": 1,
                "price": 11
              }
            ]
          }
        }
      },{
        "_links": {
          "self": { "href": "/orders/124" },
          "customer": { "href": "/customer/jen", "title": "Jen Harris <jen@internet.com>" }
        },
        "total": "20.00",
        "currency": "USD",
        "status": "processing",
        "placed": "2011-01-16",
        "_embedded": {
          "basket": {
            "_links": {
              "self": { "href": "/orders/124/basket" }
            },
            "items": [
              {
                "sku": "KLM222",
                "quantity": 1,
                "price": "9.00"
              },{
                "sku": "HHI50",
                "quantity": 1,
                "price": 11.00
              }
            ]
          }
        }
      }
    ]
  }
}
EOF;
        $expected = json_decode($JSON);
        $this->assertEquals($expected, $actual);
    }
    /**
     * @group Simple
     */
    public function testSimple()
    {
        $parent = new Hal_Resource('/dogs');
        /* Add any relevent links */
        $parent->setLink(new Hal_Link('/dogs?q={text}', 'search'));
        $dogs[1] =  new Hal_Resource('/dogs/1');
        $dogs[1]->setData(array('id' => '1', 'name' => 'tiber', 'color' => 'black'));
        $dogs[2] =  new Hal_Resource('/dogs/2',array('id' => '2', 'name' => 'sally', 'color' => 'white'));
        $dogs[3] =  new Hal_Resource('/dogs/3',array('id' => '3', 'name' => 'fido', 'color' => 'gray'));
        /* Add the embedded resources */
        foreach ($dogs as $dog) {
            $parent->setEmbedded('dog', $dog);
        }
    }
}



tests/library/Hal/ResourceTest.php
<?php
/**
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * Copyright [2012] [Robert Allen]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Hal_Link test case.
 */
class Hal_ResourceTest extends PHPUnit_Framework_TestCase
{
    public function testSetSingleLink()
    {
        $parent = new Hal_Resource('/dogs');
        $parent->setLink(new Hal_Link('/dogs?q={text}', 'search'));

        $actual = json_decode($parent);

        $JSON = <<<EOF
{
   "_links":{
      "self":{
         "href":"\/dogs"
      },
      "search":{
         "href":"\/dogs?q={text}"
      }
   }
}
EOF;

        $expected = json_decode($JSON);
        $this->assertEquals($expected, $actual);
    }

    public function testSetLinkMultiple()
    {
        $parent = new Hal_Resource('/dogs');
        $parent->setLink(new Hal_Link('/dogs?q={text}', 'search'));
        $parent->setLink(new Hal_Link('/dogs?q={text}&limit={limit}', 'search'));

        $actual = json_decode($parent);

        $JSON = <<<EOF
{
   "_links":{
      "self":{
         "href":"\/dogs"
      },
      "search":[{
         "href":"\/dogs?q={text}"
      },{
         "href":"\/dogs?q={text}&limit={limit}"
      }]
   }
}
EOF;

        $expected = json_decode($JSON);
        $this->assertEquals($expected, $actual);
    }

    public function testSetLinkMultipleSingular()
    {
        $parent = new Hal_Resource('/dogs');
        $parent->setLink(new Hal_Link('/dogs?q={text}', 'search'));
        $parent->setLink(new Hal_Link('/dogs?q={text}&limit={limit}', 'search'), true);

        $actual = json_decode($parent);

        $JSON = <<<EOF
{
   "_links":{
      "self":{
         "href":"\/dogs"
      },
      "search":{
         "href":"\/dogs?q={text}&limit={limit}"
      }
   }
}
EOF;

        $expected = json_decode($JSON);
        $this->assertEquals($expected, $actual);
    }
}


library/Hal/Link.php
<?php
/**
 *
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * Copyright [2012] [Robert Allen]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @package Hal
 */
class Hal_Link extends Hal_AbstractHal
{
    /**
     * For labeling the destination of a link with a human-readable identifier.
     * @var string
     */
    protected $_title;
    /**
     * For identifying how the target URI relates to the 'Subject Resource'.
     * The Subject Resource is the closest parent Resource element.
     *
     * @var string
     */
    protected $_rel;
    /**
     * - <b>As a resource:</b>
     *     Content embedded within a Resource element MAY be a full, partial, summary,
     *     or incorrect representation of the content available at the target URI.
     *     Applications which use HAL MAY clarify the integrity of specific embedded
     *     content via the description of the relevant @rel value.
     * - <b>As a link:</b>
     *     This attribute MAY contain a URI template. Whether or not this is the case
     *     SHOULD be indicated to clients by the @rel value.
     *
     * @var string
     */
    protected $_href;
    /**
     *
     * For distinguishing between Resource and Link elements that share the
     * same @rel value. The @name attribute SHOULD NOT be used exclusively
     * for identifying elements within a HAL representation, it is intended
     * only as a 'secondary key' to a given @rel value.
     * @var string
     */
    protected $_name;
    /**
     * For indicating what the language of the result of dereferencing the
     * link should be.
     * @var string
     */
    protected $_hreflang;
    /**
     *
     * @param string $href
     * @param string $rel
     * @param string $title
     * @param string $name
     * @param string $hreflang
     */
    public function __construct($href, $rel = 'self', $title = null, $name = null, $hreflang = null)
    {
        $this->setHref($href)
            ->setRel($rel)
            ->setName($name)
            ->setTitle($title)
            ->setHreflang($hreflang);
    }
    /**
     * @return string
     */
    public function getRel ()
    {
        return $this->_rel;
    }

    /**
     * @return string
     */
    public function getHref ()
    {
        return $this->_href;
    }
    /**
     *
     * @return string
     */
    public function getName()
    {
        return $this->_name;
    }
    /**
     * @return string
     */
    public function getTitle ()
    {
        return $this->_title;
    }

    /**
     * @return string
     */
    public function getHreflang ()
    {
        return $this->_hreflang;
    }
    /**
     * @param string $rel
     * @return Hal_Link
     */
    public function setRel ($rel)
    {
        $this->_rel = $rel;
        return $this;
    }
    /**
     * @param string $href
     * @return Hal_Link
     */
    public function setHref ($href)
    {
        $this->_href = $href;
        return $this;
    }
    /**
     *
     * @param string $name
     * @return Hal_Link
     */
    public function setName($name)
    {
        $this->_name = $name;
        return $this;
    }
    /**
     * @param string $title
     * @return Hal_Link
     */
    public function setTitle ($title)
    {
        $this->_title = $title;
        return $this;
    }

    /**
     * @param string $hreflang
     * @return Hal_Link
     */
    public function setHreflang ($hreflang)
    {
        $this->_hreflang = $hreflang;
        return $this;
    }
    /**
     *
     * @return array
     */
    public function toArray()
    {
        $link = array('href' => $this->getHref());
        if($this->getTitle()){
            $link['title'] = $this->getTitle();
        }
        if($this->getTitle()){
            $link['title'] = $this->getTitle();
        }
        if($this->getRel()){
//             $link['rel'] = $this->getRel();
        }
        if($this->getName()){
            $link['name'] = $this->getName();
        }
        if($this->getHreflang()){
            $link['hreflang'] = $this->getHreflang();
        }
        return $link;
    }
    /**
     * @return SimpleXMLElement
     */
    public function addLink(SimpleXMLElement $xml)
    {
        $this->_xml = new SimpleXMLElement('<link/>');
        $this->setXMLAttributes($this);
        return $this->_xml;
    }
}

library/Hal/AbstractHal.php
<?php
/**
 *
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * Copyright [2012] [Robert Allen]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *
 * @package Hal
 *
 *
 */
abstract class Hal_AbstractHal
{
    /**
     *
     * @var SimpleXMLElement
     */
    protected $_xml;
    /**
     * @param SimpleXMLElement $xml
     * @param Hal_Link $link
     * @return Hal_AbstractHal
     */
    public function setXMLAttributes(SimpleXMLElement $xml, Hal_Link $link)
    {
        $xml->addAttribute('href', $link->getHref());
        if($link->getRel() && $link->getRel() !== 'self'){
            $xml->addAttribute('rel', $link->getRel());
        }
        if($link->getName()){
            $xml->addAttribute('name', $link->getName());
        }
        if($link->getTitle()){
            $xml->addAttribute('title', $link->getTitle());
        }
        if($link->getHreflang()){
            $xml->addAttribute('hreflang', $link->getHreflang());
        }
        return $this;
    }
}

scripts/package.php
#!/usr/bin/env php
<?php
error_reporting(error_reporting() ^ E_DEPRECATED);
if (version_compare(PHP_VERSION, '5.3.2') >= 0) {
    error_reporting(error_reporting() ^ E_DEPRECATED);
}
date_default_timezone_set('America/Chicago');

require_once 'PEAR/PackageFileManager2.php';
PEAR::setErrorHandling(PEAR_ERROR_DIE);

/**
 * Recursively populated $GLOBALS['files']
 *
 * @param string $path The path to glob through.
 *
 * @return void
 * @uses   $GLOBALS['files']
 */
function readDirectory($path)
{
    foreach (glob($path . '/*') as $file) {
        if (!is_dir($file)) {
            $GLOBALS['files'][] = $file;
        } else {
            readDirectory($file);
        }
    }
}

$outsideDir = realpath(dirname(dirname(__FILE__)));

$version = file_get_contents($outsideDir . '/VERSION');

$api_version     = $version;
$api_state       = 'beta';

$release_version = $version;
$release_state   = 'beta';
$release_notes   = "This is an alpha release, see readme.md for examples.";

$summary     = "Hal - A PHP library for HAL generation";

$description =<<<EOF
HAL is a simple way of linking with JSON or XML.

It provides a set of conventions for expressing hyperlinks to, and embeddedness
of, related resources - the rest of a HAL document is just plain old JSON or XML.

HAL is a bit like HTML for machines, in that it is designed to drive many
different types of application. The difference is that HTML is intended for
presenting a graphical hypertext interface to a 'human actor', whereas HAL is
intended for presenting a machine hypertext interface to 'automated actors'.

This document contains a formalised specification of HAL. For a friendlier, more
 pracitcal introduction to HAL you can read this article: JSON Linking with HAL

HAL has two main components: Resources and Links.

 * https://github.com/zircote/Hal
 * http://groups.google.com/group/hal-discuss
 * http://stateless.co/hal_specification.html
 * http://blog.stateless.co/post/13296666138/json-linking-with-hal
 * http://www.mnot.net/blog/2011/11/25/linking_in_json
 * https://gist.github.com/2289546

EOF;

$package = new PEAR_PackageFileManager2();

$package->setOptions(
    array(
        'filelistgenerator'       => 'file',
        'outputdirectory'         => dirname(dirname(__FILE__)),
        'simpleoutput'            => true,
        'baseinstalldir'          => '/',
        'packagedirectory'        => $outsideDir,
        'dir_roles'               => array(
            'benchmarks'          => 'doc',
            'examples'            => 'doc',
            'library'             => 'php',
            'library/Hal'         => 'php',
            'tests'               => 'test',
        ),
        'exceptions'              => array(
            'CHANGELOG'           => 'doc',
            'readme.md'           => 'doc',
            'VERSION'             => 'doc',
            'LICENSE-2.0.txt'     => 'doc',
        ),
        'ignore'                  => array(
            'build/*',
            'package.xml',
            'build.xml',
            'scripts/*',
            '.git',
            '.gitignore',
            'tests/phpunit.xml',
            'tests/build*',
            '.project',
            '.buildpath',
            '.settings',
            '*.tgz'
        )
    )
);

$package->setPackage('Hal');
$package->setSummary($summary);
$package->setDescription($description);
$package->setChannel('pear.zircote.com');
$package->setPackageType('php');
$package->setLicense(
    'Apache 2.0',
    'http://www.apache.org/licenses/LICENSE-2.0'
);

$package->setNotes($release_notes);
$package->setReleaseVersion($release_version);
$package->setReleaseStability($release_state);
$package->setAPIVersion($api_version);
$package->setAPIStability($api_state);
/**
 * Dependencies
 */

$maintainers = array(
    array(
        'name'  => 'Robert Allen',
        'user'  => 'zircote',
        'email' => 'zircote@gmail.com',
        'role'  => 'lead',
    )
);

foreach ($maintainers as $_m) {
    $package->addMaintainer(
        $_m['role'],
        $_m['user'],
        $_m['name'],
        $_m['email']
    );
}

$files = array(); // classes and tests
readDirectory($outsideDir . '/library');
readDirectory($outsideDir . '/tests');

$base = $outsideDir . '/';

foreach ($files as $file) {

    $file = str_replace($base, '', $file);

    $package->addReplacement(
        $file,
        'package-info',
        '@name@',
        'name'
    );

    $package->addReplacement(
        $file,
        'package-info',
        '@package_version@',
        'version'
    );
}

$files = array(); // reset global
readDirectory($outsideDir . '/library');

foreach ($files as $file) {
    $file = str_replace($base, '', $file);
    $package->addInstallAs($file, str_replace('library/', '', $file));
}


$package->setPhpDep('5.2.1');

$package->setPearInstallerDep('1.7.0');
$package->generateContents();
$package->addRelease();

if (   isset($_GET['make'])
    || (isset($_SERVER['argv']) && @$_SERVER['argv'][1] == 'make')
) {
    $package->writePackageFile();
} else {
    $package->debugPackageFile();
}